### PR TITLE
Open/Closed bounds for interval arithmetic

### LINKED ISSUE #5909: Open/Closed bounds for interval arithmetic
### Is your feature request related to a problem or challenge?

Currently, we cannot differentiate `x > y` and `x >= y` while propagating comparison in [constraint propagator/solver](https://github.com/apache/arrow-datafusion/blob/main/datafusion/physical-expr/src/intervals/cp_solver.rs). 

### Describe the solution you'd like

This problem originates from the `Interval` struct in [interval arithmetic](https://github.com/apache/arrow-datafusion/blob/main/datafusion/physical-expr/src/intervals/interval_aritmetic.rs) not having a bound type. We can use an additional open/closed parameter to represent a bound. Then, related arithmetic operations are updated adhering to mathematical rules.

### Describe alternatives you've considered

-

### Additional context

-

### PR DESCRIPTION
**Pull request by berkaysynnada**
# Which issue does this PR close?

Closes #5909.

# Rationale for this change

Currently, we cannot differentiate `x > y` and `x >= y` while propagating comparison in [constraint propagator/solver](https://github.com/apache/arrow-datafusion/blob/main/datafusion/physical-expr/src/intervals/cp_solver.rs). Also `apply_operator` function that calculates the interval of the target node cannot work with `Operator::GtEq` and `Operator::LtEq`.

# What changes are included in this PR?

The `Interval` struct is modified to store bound variants. Arithmetic ops such as `lt` and `gt` are updated. `lt_eq` and `gt_eq` is added. 

# Are these changes tested?

Yes, both the interval arithmetic and symmetric hash join tests are extended.

# Are there any user-facing changes?

### CODE REVIEW

**Review by mustafasrepo — APPROVED** [review]
This Pr is passed from our internal review process. It is LGTM!. We welcome community feedback.

**Comment by ozankabak** [comment]
This is a part of the effort to make the interval library general enough to cover all use cases in the codebase, so that we can unify the impls, as was agreed on with @alamb

File: datafusion/physical-expr/src/intervals/interval_aritmetic.rs [hunk 1]
```diff
@@ -324,25 +507,24 @@ mod tests {
...
+                ))?
                 .unwrap(),
-                Interval {
-                    lower: ScalarValue::Int64(case.4),
-                    upper: ScalarValue::Int64(case.5)
-                }
+                Interval::new(
+                    IntervalBound::new(ScalarValue::Int64(case.4), true),
+                    IntervalBound::new(ScalarValue::Int64(case.5), true),
+                )
```
Comment (alamb) [inline datafusion/physical-expr/src/intervals/interval_aritmetic.rs:527]: I can't help thinking while reading this PR that all this repetition obscures somewhat the intent of the test. I wonder if there is some way to write this code like

```
            assert_eq!(
                Interval::new(
                  IntervalBound::open(case.0), IntervalBound::open(case.1)
                )
                .intersect(&Interval::new(
                  IntervalBound::open(case.2), IntervalBound::open(case.3))
                 )?
                .unwrap(),
                Interval::new(
                    IntervalBound::open(case.4),
                    IntervalBound::open(case.5 ),
                )
```

Or maybe even


```
            assert_eq!(
                Interval::both_open(case.0, case.1)
                .intersect(&Interval::both_open(case.2, case.3)))?
                Interval::both_open(case.4, case.5 ),
           )
```

Even if you just made a helper for tests, I think reducing the boilder plate would significantly improve the readability of these tests

**Review by alamb — COMMENTED** [review]
Thank you @berkaysynnada  and @mustafasrepo  -- I took a brief look through this code and it looks good to me. I left some minor style suggestions, but nothing required.

I suggest leaving the PR open for another day to gather community comments but otherwise I think it is good to merge when you are ready

File: datafusion/physical-expr/src/intervals/interval_aritmetic.rs (hunk 1 above)
Comment (ozankabak) [reply datafusion/physical-expr/src/intervals/interval_aritmetic.rs:527]:
```rust
assert_eq!(
  Interval::both_open(case.0, case.1)
  .intersect(&Interval::both_open(case.2, case.3)))?
  Interval::both_open(case.4, case.5 ),
)
```

I like this idea, let's add convenience functions `open_open`, `open_closed`, `closed_open` and `closed_closed` and use when appropriate (e.g. in tests).

**Comment by ozankabak** [comment]
> I suggest leaving the PR open for another day to gather community comments but otherwise I think it is good to merge when you are ready

Will do, thank you @alamb

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Thanks -- this is looking very nice now 👌

**Pull request merged**